auth_totp/controllers/main.py

```diff
@@ -0,0 +1,124 @@
+            if device_cookie.get('device_id') in user.trusted_device_ids.ids:
+                return response
+
+        user.generate_mfa_login_token()
+        request.session.logout(keep_db=True)
+        return http.local_redirect(
+            '/auth_totp/login',
+            query={
+                'mfa_login_token': user.mfa_login_token,
+                'redirect': request.params.get('redirect'),
+            },
+            keep_hash=True,
+        )
+
+    @http.route('/auth_totp/login', type='http', auth='none', methods=['GET'])
+    def mfa_login_form(self, *args, **kwargs):
+        return request.render('auth_totp.mfa_login', qcontext=request.params)
+
+    @http.route('/auth_totp/login', type='http', auth='none', methods=['POST'])
+    def mfa_login(self, *args, **kwargs):
+        token = request.params.get('mfa_login_token')
+        try:
+            user = request.env['res.users'].user_from_mfa_login_token(token)
+        except (MfaTokenInvalidError, MfaTokenExpiredError) as exception:
+            if isinstance(exception, MfaTokenInvalidError):
+                msg = 'Your MFA login token is not valid. Please try again.'
+            if isinstance(exception, MfaTokenExpiredError):
+                msg = 'Your MFA login token has expired. Please try again.'
+
+            return http.local_redirect(
+                '/web/login',
+                query={
+                    'redirect': request.params.get('redirect'),
+                    'error': msg,
+                },
+                keep_hash=True,
+            )
+
+        confirmation_code = request.params.get('confirmation_code')
+        if not user.sudo().validate_mfa_confirmation_code(confirmation_code):
+            return http.local_redirect(
+                '/auth_totp/login',
+                query={
+                    'redirect': request.params.get('redirect'),
+                    'error': 'Your confirmation code is not correct. Please'
+                             ' try again.',
+                    'mfa_login_token': token,
+                },
+                keep_hash=True,
+            )
+
+        # These context managers trigger a safe commit, which persists the
+        # changes right away and is needed for the auth call
+        with Environment.manage():
+            with registry(request.db).cursor() as temp_cr:
+                temp_env = Environment(temp_cr, SUPERUSER_ID, request.context)
+                temp_user = temp_env['res.users'].browse(user.id)
+                temp_user.generate_mfa_login_token(60 * 24 * 30)
+                token = temp_user.mfa_login_token
+        request.session.authenticate(request.db, user.login, token, user.id)
+
+        redirect = request.params.get('redirect')
+        if not redirect:
+            redirect = '/web'
+        response = Response(http.redirect_with_hash(redirect))
+
+        if request.params.get('remember_device'):
+            device_model = request.env['res.users.device']
```

Comment:
A cookie is data. All data construction should take place in models. Only data adaptation can take place in controllers. You are correct that not all of this logic needs to be moved over, just the stuff that defines the data that we expect elsewhere.

This also allows for someone to inherit your data model and inject values into the cookie, or completely override the operation of it altogether. As it stands here, in order to change the cookie, you have to re-implement the entire method.

More general advice - If you find yourself creating and updating records in a controller, you should probably rethink the architecture (unless you are simply passing values to a create method or something - like a CRUD method).